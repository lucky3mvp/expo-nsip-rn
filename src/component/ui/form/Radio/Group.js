/**
 * RadioGroup
 * @author asy
 */
import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import UIComponent from '../../../common/UIComponent';

export default class Group extends UIComponent {
	static defaultProps = {}
	static propTypes = {
		// 受控属性，选中的项对应的value，需配合onChange使用
		value: PropTypes.string,
		// 非受控属性，默认选中的项对应的value
		defaultValue: PropTypes.string,
		// 状态改变回调
		onChange: PropTypes.func,
		// 子元素
		children: PropTypes.node,
	}
	constructor(props) {
		super(props);

		this.state = {
			value: 'value' in props ? props.value : props.defaultValue
		}
	}
	componentWillReceiveProps(nextProps) {
		if ('value' in nextProps && nextProps.value !== this.props.value) {
			this.setState({
				value: nextProps.value
			})
		}
	}
	onChange = (checked, ele) => {
		let { onChange } = this.props;
		if (!('value' in this.props)) {
			this.setState({
				value: ele.props.value
			})
		}
		onChange instanceof Function && onChange(ele.props.value, this)
	}
	render() {
		return (
			<View style={this.style.container}>{
				React.Children.map(this.props.children, (child) => React.cloneElement(child, {
					checked: child.props.value === this.state.value,
					onChange: this.onChange
				}, child.props.children))
			}</View>
		)
	}
}

Group.baseStyle = {
	container: {
		flex: 1,
		flexDirection: 'row'
	}
}