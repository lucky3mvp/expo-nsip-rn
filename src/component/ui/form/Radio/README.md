---
title: Radio
subTitle: 单选框
---

# 单选框
### 组件描述
- 单选选择框


### 示例代码

```js
<Radio defaultChecked={false}>非受控组件：使用defaultChecked</Radio>

<Radio checked={this.state.checked2} onChange={(checked)=>{
	this.setState({
		checked2: checked
	})
}}>受控组件：使用checked</Radio>
```

## API

### props列表

属性 | 说明 | 类型 | 默认值
----|-----|------|------
| checked | 受控属性：需配合 onChange 使用更新数据 checked | boolean | - |
| defaultChecked | 非控属性：默认选中状态 | boolean | false |
| disabled | 是否禁止选择框 | boolean | false |
| iconFamily | 图标字体名 | string | - |
| iconCheckName | 选中图标名称 | string | check-square-o |
| iconUncheckName | 未选中图标名称 | string | square-o |
| onChange | 选择框变化回调，当使用 checked 属性时需配合 onChange 更新数据checked 属性 | function | - |

### 样式对象styles

属性 | 说明 | 适用类型
----|-----|------
| container | 容器样式 | View |
| icon | 图标样式 | Icon |
| text | 选择框文字样式 | Text |
| checked | 选中时的样式 | Icon |
| disabled | 禁止时的样式 | Icon，Text |


### RadioGroup

### 示例代码

```js
// Radio 需有 value 属性用于匹配选中非选中的值
<Radio.Group>
	<Radio value='english'>英语</Radio>
	<Radio value='chinese'>语文</Radio>
	<Radio value='math'>数学</Radio>
</Radio.Group>

<Radio.Group value={this.state.radioGroupValue} onChange={(value)=>{
	this.setState({
		radioGroupValue: value
	})
}}>
	<Radio value='english'>英语</Radio>
	<Radio value='chinese'>语文</Radio>
	<Radio value='math'>数学</Radio>
</Radio.Group>
```

## API

### props列表

属性 | 说明 | 类型 | 默认值
----|-----|------|------
| value | 受控属性，选中的值，需配合 onChange 使用 | stirng | - |
| defaultValue | 非控属性，默认选中的值 | string | - |
| onChange | 选择框变化回调 | function(checkedValue) | - |
| children | 选择框 | Radio | - |
