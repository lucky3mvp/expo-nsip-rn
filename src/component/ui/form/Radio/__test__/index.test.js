import React from 'react'
import Renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Radio from '../index'

describe('Radio Tests', () => {
	describe('Radio item', () => {
		it('renders correctly', () => {
			const tree = Renderer.create((
				<Radio />
			)).toJSON();
			
			expect(tree).toMatchSnapshot();
		});
	
		it('press event', () => {
			const wrapper = shallow(
				<Radio>我是一个Radio</Radio>
			);
			expect(wrapper.instance().state.checked).toBe(false);
			wrapper.simulate('press');
			expect(wrapper.instance().state.checked).toBe(true);
		});
	
		it('checked props', () => {
			let handler = jest.fn();
			const wrapper = shallow(
				<Radio checked={false} onChange={handler}>使用受控属性checked的Radio</Radio>
			);
			wrapper.simulate('press');
			expect(handler).toBeCalled();
			// 使用这种方式来模拟应该在onChange中执行的setState操作吧
			wrapper.setState({
				checked: true
			});
			expect(wrapper.instance().state.checked).toBe(true);
		});
	
		it('wrapper name', () => {
			const wrapper = shallow(
				<Radio>Radio</Radio>
			);
			expect(wrapper.name()).toMatch(/TouchableWithoutFeedback/);
		});
	});

	describe('Radio group', () => {
		let content;
		beforeEach(() => {
			content = [
				<Radio value='english' key='english'>英语</Radio>,
				<Radio value='chinese' key='chinese'>语文</Radio>,
				<Radio value='math' key='math'>数学</Radio>
			]
		});
		
		it('renders correctly', () => {
			const tree = Renderer.create((
				<Radio.Group>
					{content}
				</Radio.Group>
			)).toJSON();
			
			expect(tree).toMatchSnapshot();
		});

		it('defaultValue', () => {
			const wrapper = shallow(
				<Radio.Group defaultValue={'english'}>
					{content}
				</Radio.Group>
			);
			expect(wrapper.find(Radio)).toHaveLength(3);
			expect(wrapper.instance().state.value).toEqual('english');
		})
	});
})
