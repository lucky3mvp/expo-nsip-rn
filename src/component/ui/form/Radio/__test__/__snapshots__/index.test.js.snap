// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Tests Radio group renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Array [
            Object {
              "color": "#333",
              "fontSize": 16,
              "paddingRight": 0,
              "width": 20,
            },
            null,
            null,
          ],
          Object {},
          Object {
            "fontFamily": "FontAwesome",
          },
        ]
      }
    >
      circle-o
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "flex": 1,
            "fontSize": 14,
          },
          null,
        ]
      }
    >
      英语
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Array [
            Object {
              "color": "#333",
              "fontSize": 16,
              "paddingRight": 0,
              "width": 20,
            },
            null,
            null,
          ],
          Object {},
          Object {
            "fontFamily": "FontAwesome",
          },
        ]
      }
    >
      circle-o
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "flex": 1,
            "fontSize": 14,
          },
          null,
        ]
      }
    >
      语文
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Array [
            Object {
              "color": "#333",
              "fontSize": 16,
              "paddingRight": 0,
              "width": 20,
            },
            null,
            null,
          ],
          Object {},
          Object {
            "fontFamily": "FontAwesome",
          },
        ]
      }
    >
      circle-o
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#333",
            "flex": 1,
            "fontSize": 14,
          },
          null,
        ]
      }
    >
      数学
    </Text>
  </View>
</View>
`;

exports[`Radio Tests Radio item renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
    }
  }
  testID={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#333",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Array [
          Object {
            "color": "#333",
            "fontSize": 16,
            "paddingRight": 0,
            "width": 20,
          },
          null,
          null,
        ],
        Object {},
        Object {
          "fontFamily": "FontAwesome",
        },
      ]
    }
  >
    circle-o
  </Text>
</View>
`;
